File: backend/src/services/pptx_edit_bridge.js
/**
 * pptx-svg Node.js JSON-RPC bridge.
 *
 * Long-running process that holds PptxRenderer instances keyed by session ID.
 * Communicates with the Python wrapper via line-delimited JSON over stdin/stdout.
 *
 * Protocol:
 *   stdin  -> {"id": 1, "method": "load_pptx", "params": {"session_id": "abc", "base64": "..."}}
 *   stdout <- {"id": 1, "result": {"slide_count": 5}} | {"id": 1, "error": "..."}
 */

const { PptxRenderer, findShapeElement, getShapeTransform, emuToPx } = require("pptx-svg");
const { JSDOM } = require("jsdom");
const fs = require("fs");
const path = require("path");
const readline = require("readline");

// Session map: session_id -> { renderer, slideCount }
const sessions = new Map();

// Wasm binary — read once, reuse for all renderers
let wasmBuffer = null;

function getWasmBuffer() {
  if (!wasmBuffer) {
    const wasmPath = require.resolve("pptx-svg/dist/main.wasm");
    wasmBuffer = fs.readFileSync(wasmPath).buffer;
  }
  return wasmBuffer;
}

// --- SVG parsing to extract structured shape info ---

function parseSvgShapes(svgString) {
  const dom = new JSDOM(svgString, { contentType: "image/svg+xml" });
  const doc = dom.window.document;
  const svgEl = doc.querySelector("svg");
  if (!svgEl) return { shapes: [], slide_width_emu: 0, slide_height_emu: 0 };

  const slideCx = parseInt(svgEl.getAttribute("data-ooxml-slide-cx") || "9144000", 10);
  const slideCy = parseInt(svgEl.getAttribute("data-ooxml-slide-cy") || "6858000", 10);

  const shapeGs = doc.querySelectorAll("g[data-ooxml-shape-idx]");
  const shapes = [];

  for (const g of shapeGs) {
    const idx = parseInt(g.getAttribute("data-ooxml-shape-idx") || "-1", 10);
    const shapeType = g.getAttribute("data-ooxml-shape-type") || "?";
    const geom = g.getAttribute("data-ooxml-geom") || "";
    const fillHex = g.getAttribute("data-ooxml-fill") || "";

    // Position from data attributes
    const x = parseInt(g.getAttribute("data-ooxml-x") || "0", 10);
    const y = parseInt(g.getAttribute("data-ooxml-y") || "0", 10);
    const cx = parseInt(g.getAttribute("data-ooxml-cx") || "0", 10);
    const cy = parseInt(g.getAttribute("data-ooxml-cy") || "0", 10);
    const rot = parseInt(g.getAttribute("data-ooxml-rot") || "0", 10);

    // Extract text runs
    const textRuns = [];
    const runTspans = g.querySelectorAll("tspan[data-ooxml-run-idx]");
    const seen = new Map();
    for (const ts of runTspans) {
      const ri = ts.getAttribute("data-ooxml-run-idx");
      const paraTspan = ts.closest("tspan[data-ooxml-para-idx]");
      const pi = paraTspan ? paraTspan.getAttribute("data-ooxml-para-idx") : null;
      if (pi === null || ri === null) continue;
      const key = `${pi}:${ri}`;
      if (seen.has(key)) {
        textRuns[seen.get(key)].text += ts.textContent || "";
      } else {
        seen.set(key, textRuns.length);
        textRuns.push({ pi: parseInt(pi), ri: parseInt(ri), text: ts.textContent || "" });
      }
    }

    shapes.push({
      idx,
      shape_type: shapeType,
      geom: geom || undefined,
      fill_hex: fillHex.length === 6 ? fillHex : undefined,
      x, y, cx, cy, rot,
      text_runs: textRuns.length > 0 ? textRuns : undefined,
    });
  }

  return { shapes, slide_width_emu: slideCx, slide_height_emu: slideCy };
}

// --- RPC handlers ---

const handlers = {
  async load_pptx({ session_id, base64 }) {
    const buffer = Buffer.from(base64, "base64");
    const renderer = new PptxRenderer({ logLevel: "error" });
    await renderer.init(getWasmBuffer());
    const { slideCount } = await renderer.loadPptx(buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength));
    sessions.set(session_id, { renderer, slideCount });
    return { slide_count: slideCount };
  },

  async get_slide_info({ session_id, slide_idx }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const svgString = sess.renderer.renderSlideSvg(slide_idx);
    if (svgString.startsWith("ERROR:")) throw new Error(svgString);
    const info = parseSvgShapes(svgString);
    return { slide_idx, ...info, svg: svgString };
  },

  async get_all_slides_info({ session_id }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const slides = [];
    for (let i = 0; i < sess.slideCount; i++) {
      const svgString = sess.renderer.renderSlideSvg(i);
      if (svgString.startsWith("ERROR:")) {
        slides.push({ slide_idx: i, error: svgString });
        continue;
      }
      const info = parseSvgShapes(svgString);
      // Don't include SVG in bulk response to keep payload small
      slides.push({ slide_idx: i, ...info });
    }
    return { slide_count: sess.slideCount, slides };
  },

  async update_shape_text({ session_id, slide_idx, shape_idx, para_idx, run_idx, text }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const result = sess.renderer.updateShapeText(slide_idx, shape_idx, para_idx, run_idx, text);
    if (result.startsWith("ERROR:")) throw new Error(result);
    return { ok: true };
  },

  async update_shape_fill({ session_id, slide_idx, shape_idx, r, g, b }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const result = sess.renderer.updateShapeFill(slide_idx, shape_idx, r, g, b);
    if (result.startsWith("ERROR:")) throw new Error(result);
    return { ok: true };
  },

  async update_shape_transform({ session_id, slide_idx, shape_idx, x, y, cx, cy, rot }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const result = sess.renderer.updateShapeTransform(slide_idx, shape_idx, x, y, cx, cy, rot);
    if (result.startsWith("ERROR:")) throw new Error(result);
    return { ok: true };
  },

  async export_pptx({ session_id }) {
    const sess = sessions.get(session_id);
    if (!sess) throw new Error(`Session not found: ${session_id}`);
    const buffer = await sess.renderer.exportPptx();
    return { base64: Buffer.from(buffer).toString("base64") };
  },

  async unload({ session_id }) {
    sessions.delete(session_id);
    return { ok: true };
  },
};

// --- Main loop ---

const rl = readline.createInterface({ input: process.stdin, terminal: false });

rl.on("line", async (line) => {
  let id = null;
  try {
    const req = JSON.parse(line);
    id = req.id;
    const handler = handlers[req.method];
    if (!handler) throw new Error(`Unknown method: ${req.method}`);
    const result = await handler(req.params || {});
    process.stdout.write(JSON.stringify({ id, result }) + "\n");
  } catch (err) {
    process.stdout.write(JSON.stringify({ id, error: err.message || String(err) }) + "\n");
  }
});

rl.on("close", () => process.exit(0));

// Signal readiness
process.stdout.write(JSON.stringify({ id: null, result: "ready" }) + "\n");

File: backend/src/services/pptx_edit_bridge.py
"""Python async wrapper for the pptx-svg Node.js JSON-RPC bridge."""

import asyncio
import base64
import json
import logging
import os
import pathlib
from typing import Any

from src.constants import BRIDGE_CALL_TIMEOUT, BRIDGE_STARTUP_TIMEOUT

logger = logging.getLogger(__name__)

_BRIDGE_SCRIPT = pathlib.Path(__file__).resolve().parent / "pptx_edit_bridge.js"
_NODE_PATH = os.environ.get("NODE_PATH", "/usr/lib/node_modules")


class PptxEditBridge:
    """Manages a long-running Node.js pptx-svg bridge process."""

    def __init__(self) -> None:
        self._proc: asyncio.subprocess.Process | None = None
        self._lock = asyncio.Lock()
        self._counter = 0
        self._pending: dict[int, asyncio.Future] = {}
        self._reader_task: asyncio.Task | None = None

    async def _ensure_started(self) -> None:
        if self._proc is not None and self._proc.returncode is None:
            return
        logger.info("Starting pptx-svg bridge process")
        self._proc = await asyncio.create_subprocess_exec(
            "node",
            str(_BRIDGE_SCRIPT),
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env={**os.environ, "NODE_PATH": _NODE_PATH},
        )
        self._reader_task = asyncio.create_task(self._read_loop())
        # Wait for "ready" signal
        ready_future: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[0] = ready_future  # id=null maps to 0 in our scheme
        try:
            await asyncio.wait_for(ready_future, timeout=BRIDGE_STARTUP_TIMEOUT)
        except asyncio.TimeoutError:
            logger.error("Bridge process did not signal ready in time")
            raise RuntimeError("pptx-svg bridge startup timeout")

    async def _read_loop(self) -> None:
        assert self._proc and self._proc.stdout
        try:
            while True:
                line = await self._proc.stdout.readline()
                if not line:
                    break
                try:
                    msg = json.loads(line)
                except json.JSONDecodeError:
                    continue
                msg_id = msg.get("id")
                # null id from ready signal → use 0
                key = msg_id if msg_id is not None else 0
                future = self._pending.pop(key, None)
                if future and not future.done():
                    if "error" in msg:
                        future.set_exception(RuntimeError(msg["error"]))
                    else:
                        future.set_result(msg.get("result"))
        except asyncio.CancelledError:
            pass
        except Exception as e:
            logger.error(f"Bridge reader error: {e}")

    async def _call(self, method: str, params: dict[str, Any] | None = None) -> Any:
        async with self._lock:
            await self._ensure_started()
        assert self._proc and self._proc.stdin

        self._counter += 1
        req_id = self._counter
        req = json.dumps({"id": req_id, "method": method, "params": params or {}})

        future: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[req_id] = future
        self._proc.stdin.write((req + "\n").encode())
        await self._proc.stdin.drain()

        try:
            return await asyncio.wait_for(future, timeout=BRIDGE_CALL_TIMEOUT)
        except asyncio.TimeoutError:
            self._pending.pop(req_id, None)
            raise RuntimeError(f"Bridge call '{method}' timed out")

    async def load_pptx(self, session_id: str, pptx_bytes: bytes) -> dict:
        b64 = base64.b64encode(pptx_bytes).decode()
        return await self._call("load_pptx", {"session_id": session_id, "base64": b64})

    async def get_slide_info(self, session_id: str, slide_idx: int) -> dict:
        return await self._call("get_slide_info", {"session_id": session_id, "slide_idx": slide_idx})

    async def get_all_slides_info(self, session_id: str) -> dict:
        return await self._call("get_all_slides_info", {"session_id": session_id})

    async def update_shape_text(
        self, session_id: str, slide_idx: int, shape_idx: int,
        para_idx: int, run_idx: int, text: str,
    ) -> dict:
        return await self._call("update_shape_text", {
            "session_id": session_id, "slide_idx": slide_idx, "shape_idx": shape_idx,
            "para_idx": para_idx, "run_idx": run_idx, "text": text,
        })

    async def update_shape_fill(
        self, session_id: str, slide_idx: int, shape_idx: int,
        r: int, g: int, b: int,
    ) -> dict:
        return await self._call("update_shape_fill", {
            "session_id": session_id, "slide_idx": slide_idx, "shape_idx": shape_idx,
            "r": r, "g": g, "b": b,
        })

    async def update_shape_transform(
        self, session_id: str, slide_idx: int, shape_idx: int,
        x: int, y: int, cx: int, cy: int, rot: int,
    ) -> dict:
        return await self._call("update_shape_transform", {
            "session_id": session_id, "slide_idx": slide_idx, "shape_idx": shape_idx,
            "x": x, "y": y, "cx": cx, "cy": cy, "rot": rot,
        })

    async def export_pptx(self, session_id: str) -> bytes:
        result = await self._call("export_pptx", {"session_id": session_id})
        return base64.b64decode(result["base64"])

    async def unload(self, session_id: str) -> None:
        await self._call("unload", {"session_id": session_id})

    async def shutdown(self) -> None:
        if self._reader_task:
            self._reader_task.cancel()
        if self._proc and self._proc.returncode is None:
            self._proc.terminate()


# Singleton bridge instance
bridge = PptxEditBridge()
